### PR TITLE
Infer return value of None for correctly inferred functions with no returns

### LINKED ISSUE #485: Cannot infer empty functions
### Steps to reproduce
```python
import astroid
astroid.extract_node("""
def f():
    pass
f()
""").inferred()
```
### Current behavior
raises `StopIteration`

### Expected behavior
Returns `[const.NoneType]`

### ``python -c "from astroid import __pkginfo__; print(__pkginfo__.version)"`` output

2.0.0

This also applies to procedural functions which don't explicitly return any values.

### PR DESCRIPTION
**Pull request by nelfin**
## Steps

- [x] For new features or bug fixes, add a ChangeLog entry describing what your PR does.
- [X] Write a good description on what the PR does.

## Description

Functions with empty bodies or no return statements e.g. `def f(): pass` were returned as `Uninferable` from `infer_call_result`

## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :bug: Bug fix          |

## Related Issue

Closes #485 
Ref #663 

Pylint PR: https://github.com/PyCQA/pylint/pull/4428

### CODE REVIEW

**Comment by nelfin** [comment]
Rebased to fix conflict with changes to `Instance.getitem`

File: tests/unittest_inference.py
Comment (nelfin) [inline tests/unittest_inference.py:5320]: This test is unrelated to the inference change, but failed because `con` went from `Uninferable` to `None` (i.e. the behaviour tested was over-specified). There is some additional explanation in the commit message fixing this but the short story is that while fixing it I noticed that it didn't actually test the regression (i.e. it passed with and without the fix in 55076ca)

File: tests/unittest_inference.py
Comment (nelfin) [inline tests/unittest_inference.py:714]: These lines do nothing since `ast_nodes[3:]` is empty

File: astroid/bases.py
Comment (nelfin) [inline astroid/bases.py:336]: This fixes the `InvalidGetitem` test in in `test_invalid_subscripts`

**Review by hippo91 — CHANGES_REQUESTED** [review]
@nelfin thanks for this interesting MR! 
I have got some interrogations before merging it. Maybe @cdce8p could also give us his point of view.

File: astroid/scoped_nodes.py
Comment (cdce8p) [reply astroid/scoped_nodes.py:1664]: I would suggest `raise_any_is_abstract` 😄

File: astroid/scoped_nodes.py
Comment (cdce8p) [reply astroid/scoped_nodes.py:1752]: I agree, it should probably be `pass_is_abstract=True`. My thinking was that a function with just `pass` should probably not be inferred as returning `None`.

**Review by cdce8p — COMMENTED** [review]
All in all this looks like a reasonable change. Thanks for also opening https://github.com/PyCQA/pylint/pull/4428 to fix the pylint tests!

I ran some tests and didn't notice anything unusual, so should be ok to merge once all discussions are resolved.

File: tests/unittest_inference.py
Comment (nelfin) [reply tests/unittest_inference.py:714]: Since we loop through all the nodes below:

```python
for node in ast_nodes:
    self.assertRaises(InferenceError, next, node.infer())
```

File: tests/unittest_scoped_nodes.py
Comment (nelfin) [reply tests/unittest_scoped_nodes.py:518]: Sure, I'll add a comment explaining this in the code.

File: tests/unittest_inference.py
Comment (nelfin) [reply tests/unittest_inference.py:5320]: How about I move it to another MR?

> Thus the last line could be assert isinstance(next(node.infer()), Const) and assert (next(node.infer()).value == None). Am i wrong?

You're not wrong per se (though it would be `BoundMethod __enter__ of DummyContext` if I'm reading it correctly), but I didn't change that for the same reason that I detailed above: the test is *actually* testing (in my opinion), that inference of these nodes does not crash, not what you're actually inferring them as. If you over-specify the behaviour of tests, they have a tendency to be "brittle", or break when unrelated changes occur, just like this test in the first place.

File: astroid/scoped_nodes.py
Comment (nelfin) [reply astroid/scoped_nodes.py:1752]: I've updated this, I'll need to update the pylint MR since I relied on `pass_is_abstract=False`

File: tests/unittest_scoped_nodes.py
Comment (nelfin) [reply tests/unittest_scoped_nodes.py:518]: I've forgotten some of the reasoning behind this, but I'm fairly sure it was about ignoring "abstract" functions, since the change around implicit return `None` caused `pylint` to fail on the `astroid` source code for basically this reason: `NodeNG` has a stub `_infer` method that everything else overloads. I can't remember what this test was *supposed* to look like though, and it passes now so there's no real point to it until I can remember what it meant to look like (like multiple inferred return values for each subclass etc.) so I'll remove it.

File: astroid/bases.py
Comment (nelfin) [reply astroid/bases.py:337]: I'm trying to match the style of a few other inference functions: when it is known that code would raise a `Value/Index/TypeError` these exceptions signal that, for inference brains, the usual continuation is to raise `UseInferenceDefault` from that as a signal that that particular brain should be ignored. Here I'm trying to follow the same convention, if some caller calls `getitem` with a function that doesn't match the required signature then we know that Python would raise a `TypeError` at this point and so that there's no reasonable inference to return from this function. We could raise `InferenceError` I suppose, I was just trying to be *more* specific.

Comment (nelfin) [reply astroid/bases.py:337]: e.g. see `ClassDef.getitem` for an existing example: https://github.com/PyCQA/astroid/blob/d66e3d113c0b2cb9283f3518e5a494d139c6779a/astroid/scoped_nodes.py#L2661-L2674

File: tests/unittest_inference.py
Comment (hippo91) [reply tests/unittest_inference.py:5353]: @nelfin you are right. Maybe you can add a docstring to detail that it test the non-crash behavior and not the result of the inference. What do you think about it?

**Comment by hippo91** [comment]
> All in all this looks like a reasonable change. Thanks for also opening [PyCQA/pylint#4428](https://github.com/PyCQA/pylint/pull/4428) to fix the pylint tests!
> 
> I ran some tests and didn't notice anything unusual, so should be ok to merge once all discussions are resolved.

Thanks @cdce8p !

File: astroid/bases.py
Comment (hippo91) [reply astroid/bases.py:337]: It definitely makes sense. Thanks for the explanation!

File: tests/unittest_inference.py
Comment (hippo91) [reply tests/unittest_inference.py:714]: :+1:

**Review by hippo91 — APPROVED** [review]
@nelfin thanks for your answers! It is all good for me. Just left a comment regarding adding a comment but feel free to add it or not.

**Comment by nelfin** [comment]
> @nelfin thanks for your answers! It is all good for me. Just left a comment regarding adding a comment but feel free to add it or not.

I updated the comment with a more in-depth explanation

File: tests/unittest_inference.py
Comment (hippo91) [reply tests/unittest_inference.py]: :+1:

**Comment by hippo91** [comment]
Thanks @nelfin !

**Pull request merged**
